### PR TITLE
[fix][broker] fix ModularLoadManagerImpl always delete active bundle-data. sec ver.

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
### Motivation

fix https://github.com/apache/pulsar/issues/20544

this problem will cause LoadManager's loadSheding and BundleSplitTask not functioning. because of the lost of real cluster-wide bundle-data in LoadManager.

master code logic in ModularLoadManagerImpl will try to clean up bundle-data if the namespace bundle is splitted.

but the current logic will cause amost all bundle-datas in the cluster to be deleted, each time the updateBundleData method triggered.

### Modifications
move bundle-data cleanup logic after iterator all broker to get cluster wide activeBundles in updateBundleData method.

### Verifying this change

add unittest to check if the broker number is more than one.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java [hunk 1]
```diff
@@ -603,6 +592,16 @@ private void updateBundleData() {
                 LoadManagerShared.fillNamespaceToBundlesMap(preallocatedBundleData.keySet(), namespaceToBundleRange);
             }
         }
+
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:595]: Can you explain why this reordering can resolve the issue?

Comment (lifepuzzlefun) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:595]: Sure. 
If we need to figure out which bundle is not active, we should get cluster wide all bundle data, right ? To get all cluster wide bundle data, we need iterate all brokerData to collect this. 
Current we only iterate the first brokerData and check active bundles. so the all the bundleData which is not owned by the first broker will be cleaned at the first broker check.

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:595]: Sorry, maybe I missed some logic here.

But, In this func, I don't see `bundleData` and `activeBundles` are updated between these lines. https://github.com/apache/pulsar/pull/20620/files#diff-642d3e26ddad51db8fb736b0f5519f4c929814a4f69d85e16829d19b885be0ebR562-R594 .
So, I am curious to know how this reordering can be effective.

 
>  To get all cluster wide bundle data, we need iterate all brokerData to collect this.
> Current we only iterate the first brokerData and check active bundles. so the all the bundleData which is not owned by the first broker will be cleaned at the first broker check.


Can you provide the code lines which iterate all brokerData and fill `activeBundles` and `bundleData`, other than here?
https://github.com/apache/pulsar/pull/20620/files#diff-642d3e26ddad51db8fb736b0f5519f4c929814a4f69d85e16829d19b885be0ebR549-R558

**Comment by lifepuzzlefun** [comment]
before: 64205ec7-d839-4e51-9a42-233b64b6f6b9

after: 
AsvF7Xa1KU

**Comment by heesung-sn** [comment]
Oh. I mis-read the change. I got it now. This removal has been moved out of the loop, `for (Map.Entry<String, BrokerData> brokerEntry : loadData.getBrokerData().entrySet())`

```
            //Remove not active bundle from loadData
            for (String bundle : bundleData.keySet()) {
                if (!activeBundles.contains(bundle)){
                    bundleData.remove(bundle);
                    if (pulsar.getLeaderElectionService().isLeader()){
                        deleteBundleDataFromMetadataStore(bundle);
                    }
                }
            }
```

LGTM.

**Review by heesung-sn — APPROVED** [review]

**Comment by lifepuzzlefun** [comment]
@mattisonchao @Demogorgon314 please take a look, thank you :wave:

**Comment by lifepuzzlefun** [comment]
> Oh. I mis-read the change. I got it now. This removal has been moved out of the loop, `for (Map.Entry<String, BrokerData> brokerEntry : loadData.getBrokerData().entrySet())`
> 
> ```
>             //Remove not active bundle from loadData
>             for (String bundle : bundleData.keySet()) {
>                 if (!activeBundles.contains(bundle)){
>                     bundleData.remove(bundle);
>                     if (pulsar.getLeaderElectionService().isLeader()){
>                         deleteBundleDataFromMetadataStore(bundle);
>                     }
>                 }
>             }
> ```
> 
> LGTM.

Thanks for review this fix, I'll try to explain myself more clearly next time :smile:

**Comment by lifepuzzlefun** [comment]
@codelipenghui @gaozhangmin @Jason918 please take a look, thank you 👋

**Comment by lifepuzzlefun** [comment]
@Demogorgon314 could you help review this pr?

**Review by Demogorgon314 — APPROVED** [review]
LGTM.

**Comment by lifepuzzlefun** [comment]
@Technoboy- hi, this is a fix in ModularLoadManager, hope can be in 2.11.2 can you check this pr ? : - )

**Review by aloyszhang — APPROVED** [review]

**Comment by aloyszhang** [comment]
Nice cath, LGTM.

**Comment by aloyszhang** [comment]
@lifepuzzlefun plz resolve the conflict

**Comment by lifepuzzlefun** [comment]
> @lifepuzzlefun plz resolve the conflict

Thanks for review. Conflict has been resolved. : -)

**Pull request merged**